Balance for HT process

TomFischer opened the pull request:
PR implements getFlux in the in the class HTProcess and the local assembler for the HT process. The new feature is checked by two new test cases.

endJunction commented on ProcessLib/HT/HTFEM.h:
```diff
@@ -88,6 +88,76 @@ class HTFEM : public HTLocalAssemblerInterface
+        fe.computeShapeFunctions(pnt_local_coords.getCoords(), shape_matrices,
+                                 GlobalDim, false);
+        std::vector<double> flux;
+        flux.resize(3);
```
One could move it down to near the return statement.
I'd say, that returning a Eigen::Vector3d is safer, then the std::vector...

endJunction commented on ProcessLib/HT/HTFEM.h:
```diff
@@ -88,6 +88,76 @@ class HTFEM : public HTLocalAssemblerInterface
+        NumLib::shapeFunctionInterpolate(local_x, shape_matrices.N, T_int_pt, p_int_pt);
+
+        // set the values into array vars
+        vars[static_cast<int>(MaterialLib::Fluid::PropertyVariableType::T)] =
```
Would it make sense to pass the vars[T] directly to shapeFunctionInterpolate call?
```c++
NumLib::shapeFunctionInterpolate(local_x, shape_matrices.N, vars[static_cast<int>(MaterialLib::Fluid::PropertyVariableType::T)], vars[static_cast<int>(MaterialLib::Fluid::PropertyVariableType::p)]);
```

endJunction commented on ProcessLib/HT/HTFEM.h:
```diff
@@ -88,6 +88,76 @@ class HTFEM : public HTLocalAssemblerInterface
+                    MaterialLib::Fluid::FluidPropertyType::Density, vars);
+            auto const b = this->_material_properties.specific_body_force;
+            // here it is assumed that the vector b is directed 'downwards'
+            q += K_over_mu * rho_w * b;
```
I'm not getting the intentions of the comment. What is 'downwards'? Maybe it's copied from some other context...

endJunction commented on ProcessLib/HT/HTProcess.cpp:
```diff
@@ -229,5 +240,59 @@ void HTProcess::setCoupledSolutionsOfPreviousTimeStep()
+        DBUG("This is the thermal part of the staggered HTProcess.");
+        return;
+    }
+    if (_balance_mesh)  // computing the balance is optional
```
```c++
if (!_balance_mesh)
{
    return;
}
```

endJunction commented on ProcessLib/HT/HTProcess.cpp:
```diff
@@ -229,5 +240,59 @@ void HTProcess::setCoupledSolutionsOfPreviousTimeStep()
+        auto* const balance_pv = MeshLib::getOrCreateMeshProperty<double>(
+            *_balance_mesh, _balance_pv_name, MeshLib::MeshItemType::Cell, 1);
+        // initialise the PropertyVector pv with zero values
+        std::fill(balance_pv->begin(), balance_pv->end(), 0.0);
```
Not sure here;
If the properties are set, then use NaN for initialization. If the properties are '+='-ed, then setting them 0 is of course necessary.

TomFischer replied:
The properties are  '+='-ed.

chleh commented on ProcessLib/GroundwaterFlow/GroundwaterFlowFEM.h:
```diff
@@ -105,7 +105,7 @@ class LocalAssemblerData : public GroundwaterFlowLocalAssemblerInterface
     /// Computes the flux in the point \c p_local_coords that is given in local
     /// coordinates using the values from \c local_x.
     // TODO add time dependency
-    std::vector<double> getFlux(
```
~~Isn't the flux 2D in two dimentsions?~~

chleh commented on ProcessLib/HT/HTFEM.h:
```diff
@@ -88,6 +88,73 @@ class HTFEM : public HTLocalAssemblerInterface
 
+    /// Computes the flux in the point \c pnt_local_coords that is given in
+    /// local coordinates using the values from \c local_x.
+    // TODO add time dependency
```
Yep, that would be nice (for the sake of completeness). Could you please add it now?

chleh commented on ProcessLib/HT/HTFEM.h:
```diff
@@ -88,6 +88,73 @@ class HTFEM : public HTLocalAssemblerInterface
+        auto const K =
+            this->_material_properties.porous_media_properties
+                .getIntrinsicPermeability(t, pos)
+                .getValue(t, pos, 0.0,
```
Is the `0.0` correct here?

TomFischer replied:
Yes, I think so.

chleh commented on ProcessLib/HT/HTFEM.h:
```diff
@@ -88,6 +88,73 @@ class HTFEM : public HTLocalAssemblerInterface
+        }
+
+        Eigen::Vector3d flux;
+        flux.head(GlobalDim) = q;
```
Maybe here the templated version `head>GlobalDim>()` would be a tiny bit better.
And I see: That's why flux is always 3D.

endJunction replied:
Regarding the flux being 3D always: I thought it would be easier in general to be it of constant size, like the node coordinates.

`flux.head<GlobalDim>()` would be better, that's true!

chleh commented on ProcessLib/HT/HTFEM.h:
```diff
@@ -88,6 +88,73 @@ class HTFEM : public HTLocalAssemblerInterface
+    /// Computes the flux in the point \c pnt_local_coords that is given in
+    /// local coordinates using the values from \c local_x.
+    // TODO add time dependency
+    Eigen::Vector3d getFlux(
```
For the HT process the problem is that there are potentially two different interesting fluxes:
The mass/volume flux and the heat flux.
Maybe there should be different names for them?

endJunction replied:
Or one could pass the variable_id to getFlux?

TomFischer replied:
You are right, there are potentially two different interesting fluxes. Maybe other processes have also interesting quantities fluxes. For this reason we would need rather general names. I opened an issue for discussion and for the time being I would keep the implementation of this PR. Is this okay for you?

chleh replied:
OK.

TomFischer commented:
Added the time to all getFlux and the balance functions, see commit a823565. For the second bigger comment I added the issue https://github.com/ufz/ogs/issues/2134 for discussion.
